DataStructures/AATree/AATree.cs

```
@@ -0,0 +1,552 @@
+                result.Add(node.Key);
+            };
+
+            postOrderWalk(root);
+            return result;
+        }
+
+        /// <summary>
+        /// Checks if the tree is a valid AA Tree.
+        /// </summary>
+        /// <returns>true if tree is a valid AA Tree, false otherwise.</returns>
+        public bool Validate()
+        {
+            return ValidateTree(root);
+        }
+
+        /// <summary>
+        /// Recursive function to add an element to the tree.
+        /// </summary>
+        /// <param name="key">The element to add.</param>
+        /// <param name="node">The node to search for a empty spot.</param>
+        /// <returns>The node with the added element.</returns>
+        /// <exception cref="ArgumentException">Thrown if key is already in the tree.</exception>
+        private AATreeNode<TKey> Add(TKey key, AATreeNode<TKey>? node)
+        {
+            if (node == null)
+            {
+                return new AATreeNode<TKey>(key, 1);
+            }
+            else if (comparer.Compare(key, node.Key) < 0)
+            {
+                node.Left = Add(key, node.Left);
+            }
+            else if (comparer.Compare(key, node.Key) > 0)
+            {
+                node.Right = Add(key, node.Right);
+            }
+            else
+            {
+                throw new ArgumentException($"Key \"{key}\" already in tree!", nameof(key));
+            }
+
+            return Split(Skew(node)) !;
+        }
+
+        /// <summary>
+        /// Recursive function to remove an element from the tree.
+        /// </summary>
+        /// <param name="key">The element to remove.</param>
+        /// <param name="node">The node to search from.</param>
+        /// <returns>The node with the specified element removed.</returns>
+        private AATreeNode<TKey>? Remove(TKey key, AATreeNode<TKey>? node)
+        {
+            if (node == null)
+            {
+                return node;
```

When is this executed? This is not a good idea to return null, should probably throw an exception of some kind